src/coreclr/nativeaot/Runtime/amd64/ExceptionHandling.asm

```diff
@@ -539,18 +539,26 @@ endif
         ;; Find the shadow stack pointer for the frame we are going to restore to.
         ;; The SSP we search is pointing to the return address of the frame represented
         ;; by the passed in context. So we search for the instruction pointer from 
-        ;; the context or for RhpThrowHwEx2, which is there in a case of a hardware fault,
-        ;; and pop SSP to one slot up from there.
+        ;; the context and pop SSP to one slot up from there.
+        ;; When we have a hardware fault and the catch is in the same method, the IP may not be on the stack,
+        ;; then we look for RhpThrowHwEx2.
         mov     r8, RhpThrowHwEx2
-        xor     r11, r11
-   @@:  inc     r11
-        cmp     [r9 + r11 * 8 - 8], r8
-        je      SSP_OK
-        cmp     [r9 + r11 * 8 - 8], r10
-        jne     @b
-
+        mov     r11, 2                   ;; we will be popping at least 2 items for the exception dispatch
+   @@:  cmp     [r9 + r11 * 8 - 8], r10  ;; check for the IP first
+        je      FoundIP
+        cmp     [r9 + r11 * 8 - 16], r8  ;; in case IP was not there, check for RhpThrowHwEx2 at -16
+        je      FoundRhpThrowHwEx2
+        inc     r11
+        jmp     @b
+
+FoundRhpThrowHwEx2:
```

Comment:
I am not sure I understand why is the SSP increment different for the RhpThrowHwEx2 / real IP case. We don't insert any frames to the stack, so it seems the increment should be the same.

Reply:
HwEx is on the stack even when IP is present, but when IP is there, I do not want to find the HwEx. 

Thus I want to find HwEx after intentionally overshooting by 1 item.

Reply:
Basically - look for IP at -8 and for HwEx at -16. So that if IP is present I will find it first, even though HwEx is closer to the top.

Reply:
Ah, makes sense, I've missed the difference in the offset in the two cmp instructions.

I still wonder if this trick is correct in all cases. Let's assume that we have a catch handler that's couple of frames above the hardware exception frame. Something like:
```
Frames on the shadow stack:
A -> this throws the HW exception, so we will have the RhpThrowHwEx2 on the shadow stack here
B
C
D -> this frame contains the catch, so we will be resuming here
```
It seems that the code you have would first find the RhpThrowHwEx2 and update the SSP based on that while it should actually update it to the frame D above.

Reply:
right. it does not work if fault happens at deeper levels.

Reply:
I also think it is possible to construct a recursive case where the faulting IP could be present on the shadow stack multiple times, while the lowermost occurrence will not match the catch/continuation.

Reply:
We may need to count the number of pops/unwinds between the fault and the catch to pop SSP precisely.

Reply:
I've looked at how windows deals with SSP during `RtlUnwind`. I have looked at it in the past and I have believed it was scanning for the resume IP on the shadow stack at the end, but I was wrong. That code is there, but only as a mean to verify that the resume IP is on the shadow stack. The `RtlVirtualUnwind` actually unwinds the SSP if the context contains `XSTATE` with the `XSTATE_MASK_CET_U` and the `RtlUnwind` creates the context it uses during unwind with that set.
So it seems to be that the best way to handle the problem would be to do a similar thing. We could add SSP to the `REGDISPLAY` and let the `CoffNativeCodeManager::UnwindStackFrame` increment it (and also increment it during the manual unwinding in the two cases where you've added the unwindCount stuff in the last change). I think it would be cleaner to keep the SSP instead of the unwind count, unless there is some gotcha.

Reply:
For coreclr, it would probably be more involved due to the problem with starting at an arbitrary explicit frame that I've mentioned. Maybe we'll need to store SSP in those frames too, but I am not sure yet.

Reply:
Oh, and I've just found something I didn't know - the incsspq instruction can move SSP by max 255 slots only, it uses only the lowest 8 bits from the argument register for some reason. So we will need to take care of that when moving the SSP ourselves.

Reply:
> I think it would be cleaner to keep the SSP instead of the unwind count, unless there is some gotcha.

Thinking about it again, using the unwind count may actually solve the problem with unwind not knowing the initial SSP on coreclr. So maybe the way you did it would be better. I am not sure yet though.

Reply:
Unwinding SSP is basically incrementing it every time a call is unwound. It seems like just keeping the count of call unwinds is a rough equivalent. 
Counting unwinds seems simpler though, when considering multiplatform (eventually will need to do something for Linux and arm64), dealing with XSTATE, machines that do not support CET or do not have it enabled. 
Also, as you mentioned, the count is a diff, that can be applied later - when we know the actual SSP value.

Reply:
>  the incsspq instruction can move SSP by max 255 slots only, it uses only the lowest 8 bits from the argument register for some reason

That is a weird limitation. 
We can work with that, but it is a bit of an unexpected nuisance.

Reply:
>  And NativeAOT has the problem with the RhpThrowHwEx2 replacing the faulting location IP.

The root cause for that is that we pretend there was a call to `RhpThrowHw` at the faulting location by modifying the context. 

```c
    if (translateToManagedException)
    {
        *arg0Reg = faultCode;
        *arg1Reg = faultingIP;
        palContext->SetIp(PCODEToPINSTR((PCODE)&RhpThrowHwEx));

        return EXCEPTION_CONTINUE_EXECUTION;
    }
```

But there is no actual call made, thus the "fake" call is not recorded in SSP.

I wonder how CoreCLR does not have a similar problem.

Reply:
When the fault and catch are in the same method, then there is no call that would push the faulting location onto SSP, so we cant find it.

Reply:
> I wonder how CoreCLR does not have a similar problem.

We don't do it this way on coreclr - we call the exception handling code from the vectored exception handler directly. So the stack contains the fault address, then the vectored exception handler code, then the managed EH frames.

Reply:
I've actually changed coreclr to that model two years ago, it was originally also doing a similar thing as the current nativeaot. See #70428.

Reply:
It would be nice to switch naot to the same plan as regular CoreCLR. It would improve diagnosability of hardware exceptions in naor. With the current scheme, it is not possible to get to the exception location using `.cxr` windbg command that makes hardware exceptions pain to debug.

Reply:
@VSadov I have found there is one more complication here. When exception is thrown from a call chain of a catch handler and that exception escapes the handler. Exception unwind will detect that it has escaped the handler and "skip" to the context where the previous exception (the one that called the catch handler) has found the handler and then move on from there. This skipping further complicates the SSP updating via the unwind count.
I am starting to think that it would be the best to capture the SSP at the RhThrowEx / RhThrowHwEx, store it in the REGDISPLAY and then increment it the same way as you did the unwind count. That way, when skipping to the other context, the SSP would get the correct value from that context and then move on.

Reply:
> It would be nice to switch naot to the same plan as regular CoreCLR.

I did not know we do something different on CoreCLR.

Switching to the CoreCLR model would "fix" the simple case as we would always find the faulting IP on SSP. Although the more exotic recursive cases where the lowermost occurrence is not the right one might still exist and need to be addressed.  

If the switch to different model of HW exception handling is a simple change, I'd prefer to do that first and then deal with rare cases separately. 
Although I suspect the change is not trivial. The part where the vectored exception handler code will end up in the middle of the managed stack makes me a bit worried.

Reply:
> @VSadov I have found there is one more complication here. When exception is thrown from a call chain of a catch handler and that exception escapes the handler. Exception unwind will detect that it has escaped the handler and "skip" to the context where the previous exception (the one that called the catch handler) has found the handler and then move on from there. This skipping further complicates the SSP updating via the unwind count.
I am starting to think that it would be the best to capture the SSP at the RhThrowEx / RhThrowHwEx, store it in the REGDISPLAY and then increment it the same way as you did the unwind count. That way, when skipping to the other context, the SSP would get the correct value from that context and then move on.

Yes, it seems like in this case we would need to add unwind counts. 
Maybe storing the unwind counts or the actual SSP value on REGDISPLAY could make this simpler.

I am mostly just trying to keep this abstract enough that it works regardless of how the shadow stacks are implemented or whether shadow stacks are implemented at all. 
That is what attracts me in "unwind count" - it is something that makes sense on every platform and regardless of shadow stacks. So far it did not require writing any platform-specific code, except the place which uses the count to pop the SSP, which needs to be platform specific anyways.

Reply:
For coreclr, I've added SSP to the REGDISPLAY and it works nicely. On coreclr, the unwind count is not something I can use without also having the initial SSP and it seemed unnecessary to store both of them. And on other platforms, the unwind count seems quite useless to me.

Reply:
I will try SSP on the REGDISPLAY approach.

it will be AMD64 - specific.

I assume it records the value of the SSP when exception is thrown: 
- 0 if SSP does not read, or we do not care - i.e. on Linux.

otherwise it will contain a pointer to
- IP of the callsite for the throw helper (basically SSP when in the throw helper)
- IP of the callsite to the method that faulted (SSP when in the HW throw helper)

Then we +8 the value every time we unwind REGDISPLAY out of a call.
(We do not need to +8 if it is 0)

When we switch to a different REGDISPLAY, it will have the value consistent with its state, so no need to adjust.

When we need to pop to the catch continuation, we keep popping (in no more than 256 increments?) the actual SSP until it is popped to the expected value.